modules/ROOT/pages/dataweave-type-system.adoc

```diff
@@ -0,0 +1,291 @@
+= Type System
+ifndef::env-site,env-github[]
+include::_attributes.adoc[]
+endif::[]
+
+DataWeave 2.0 introduced a new type system on the language. To understand how values for these types are defined see xref:dataweave-types.adoc[Value Constructs for DataWeave Types] documentation.
+
+A type system defines a set of constraints to a set of constructs such as:
+
+* Variables
+* Function parameters
+
+These constraints are used in the type checking phase where DataWeave ensures that the values assigned
+to variables or the arguments for a function call respects its constraints.
+
+For example:
+
+----
+var userName: String = "John"
+----
+
+The variable definition for `userName` has the constraint of assigning a value of type String.
+If some other type value is assigned, a type checking error raises.
+
+----
+fun toUser(id: Number, userName: String): String = "you called the function toUser!"
+----
+
+The constraints to call the function `toUser` are that the first argument has to be of type Number, the second one of type String (or to be coercible to those types) and the result of the function has to be of type String.
+
+----
+toUser(123, "John")
+toUser("123", true)
+----
+
+Both calls don't throw an exception. In the case of the second function call, there is an autocoercion of the String `"123"` to the Number value `123`, and the Boolean `true` to the String value `"true"` to being able to call the function. +
+An example of a call that throws a type checking error is `toUser("a 12", "John")` because the String `"a 12"` isn't coercible to a Number. +
+
+DataWeave also uses a global type inference algorithm to validate your code even if no type is specified as
+constraints.
+
+Note that the use of constraints are optional but this can be useful in big scripts or multiple scripts where the type system could help you avoid bugs in your DataWeave logic by running the type checking algorithm.
+
+The types supported by the DataWeave type system are divided into three categories:
+
+* Simple Types
+* Composite Types
+* Complex Types
+
+== Simple Types
+
+Simple types represent values such as String, Boolean, etc. These values are atomic, they are not composed by other values. The list of simple types are:
+
+* String
+* Boolean
+* Number
+* Regex
+* Null
+* Temporal: Date, DateTime, LocalDateTime, LocalTime, Time, Period
+//Key, KeyValuePairType, NameType, NameValuePairType, NamespaceType, RangeType , ReferenceType, TypeParameter, TypeType, UriType
+
+=== Null
+
+Null is a type of only one value, which means that `null` can not be assigned to the type String or any other type except for the type Null.
+
+For example, the repeat function in the String Module has the signature `repeat(String, Number): String`, which means that only accepts as a first parameter a value of type String, and as a second parameter a value of type Number. The function returns a value of type String.
+
+----
```

Comment:
Is there a way to indicate that all these examples are DW code? I cannot remember.

Reply:
@dukesphere could you confirm?